### PR TITLE
Binary, Octal, and Hex conversions between [positive], [Z], [N], [nat] and [string]

### PR DESCRIPTION
**Pull request by JasonGross**
@davidben pointed out to me that Coq doesn't have conversations between numbers and strings.  This adds such conversions for `positive`, `N`, `nat`, and `Z` in binary, octal, and hex.

Edit: <strike>This is on top of #6596</strike> (not anymore)

- [x] Entry added in CHANGES.

### CODE REVIEW

File: theories/Strings/Hex.v
Comment (herbelin) [inline theories/Strings/Hex.v]: I would vote for not indenting at each new `if` here (and idem in `Octal.v`.

Comment (JasonGross) [reply theories/Strings/Hex.v]: Updated

**Comment by JasonGross** [comment]
Updated to add conversion with `nat`, and to not depend on #6596, which seems blocked on having more features around `Scheme`.

**Comment by maximedenes** [comment]
@letouzey Would you mind having a look at this one?

**Comment by maximedenes** [comment]
@JasonGross we can try to get a review from @letouzey, but I fear we haven't let him enough time to review this PR (the last update was done two days before the freeze).

**Comment by JasonGross** [comment]
@maximedenes the latest updated fixed the merge conflict in `CHANGES`, replaced `ascii_beq` with `ascii_dec`, `a || b` with a local notation `if a then true else if b then true else false` (so that it would work with `ascii_dec`), and added 
```diff
+Definition bin_string_of_nat (n : nat) : string
+  := bin_string_of_N (N.of_nat n).
+Definition nat_of_bin_string (s : string) : nat
+  := N.to_nat (N_of_bin_string s).
+Lemma nat_of_bin_string_of_nat (n : nat)
+  : nat_of_bin_string (bin_string_of_nat n)
+    = n.
+Proof.
+  cbv [nat_of_bin_string bin_string_of_nat];
+    rewrite N_of_bin_string_of_N, Nnat.Nat2N.id; reflexivity.
+Qed.
+Example bin_string_of_nat_0 : bin_string_of_nat 0 = "0b0" := eq_refl.
```
and similarly with `s/bin/oct/` and `s/bin/hex/`.  The rest of the PR has been unchanged for over a month.

**Comment by maximedenes** [comment]
@JasonGross I see, sorry about that. Maybe we can still hope to get a review from @letouzey.

File: theories/Strings/Hex.v
Comment (letouzey) [inline theories/Strings/Hex.v]: Come on ! That's `if a then true else b`. And it already exists in `Bool` (notation `|||` in scope `lazy_bool_scope`).

**Comment by letouzey** [comment]
Concerning the use of `ascii_dec` instead of #6596, why not (as a temporary step). See also my own proposal #6948.

File: theories/Strings/Hex.v
Comment (JasonGross) [reply theories/Strings/Hex.v]: No, I believe `if a then true else b` does not pass type inference, because `b` is of type `{_} + {_}` while `true` is of type `bool`.

**Comment by JasonGross** [comment]
>  See also my own proposal #6948.

Sure, I'm happy to backtrack and put this on top of #6948 if that gets merged first.

**Comment by letouzey** [comment]
Anyway, I'm quite unconvinced that these files deserve to enter the Standard Library. It looks to me as a *very* specific need, that could well live in an external opam package. Ok, I sponsored recently a bunch of files concerning decimal representation, one of them even entered `theories/Init`, but that's because we plan to make them be used by the standard parser / printer for numbers. Even if we extend that someday to binary, hex or octal, that will *not* go through Coq `string`, so your files won't be directly concerned. Ok, there is now `DecimalString.v` which is related to this PR, but for base 10, and I think it will have a larger audience than your PR files (and moreover it was pretty trivial thanks to the other stuff about `Decimal`).

**Comment by letouzey** [comment]
Now, concerning the general shape of this PR:
  - I don't like the name `Binary.v`, which is way too broad, even if it is meant to be used as `String.Binary`. Maybe `ToBinary` ? Or `BinaryConv` ? Or `BinaryString` to come along the already existing `DecimalString` ? Of course, `Hex.v` and `Octal.v` would have to be adapted accordingly.
  - In general, the function names aren't great. Why not shorten them (e.g. `to_Z` or `of_Z`) and let the module-qualified form provide long, unambiguous names when needed (e.g. `BinaryString.to_Z`).
 - The names finishing with quote (e.g. `bin_string_of_pos'` are quite ugly. Moreover they can be just as useful as their "unquoted" counterpart, since one may well be interested in converting *without" the "0b" or "0x" or "Oo" prefix. So I suggest to rather place them in submodules (for instance named `Raw`) and use nicer names.
 - Globally, the current indentation of these files is quite awful.

**Review by letouzey — CHANGES_REQUESTED** [review]
See my various comments. Overall, I'm not really in favor of merging this PR, even after revision. But I'm not strongly opposed to it either.

File: theories/Strings/Binary.v
Comment (letouzey) [inline theories/Strings/Binary.v]: That's called `N.double rest`

File: theories/Strings/Binary.v
Comment (letouzey) [inline theories/Strings/Binary.v]: And this one is `N.succ_double rest`.

File: theories/Strings/Binary.v
Comment (letouzey) [inline theories/Strings/Binary.v]: I'm not fond of pushing this `Pos.app` in `BinPosDef` just for the statement of this lemma.

File: theories/Strings/Hex.v
Comment (letouzey) [inline theories/Strings/Hex.v]: This enumeration of all these cases is really heavy. Why not shorten it to something like like
```
if ascii_dec ch "0" then 16*rest
else if ascii_dec ch "1" then 16*rest+1
else ...
```
This could even be turned into an `ascii_to_hexdigit` subfunction...

File: theories/Strings/Octal.v
Comment (letouzey) [inline theories/Strings/Octal.v]: Same remark as for `Hex.v`

File: theories/Strings/Hex.v
Comment (letouzey) [reply theories/Strings/Hex.v]: Indeed, my bad. One more reason to favor something like #6948 instead !

Comment (herbelin) [reply theories/Strings/Hex.v]: Why not `Local Notation "a || b" := (match a with true => true | false => b end).`?

Comment (JasonGross) [reply theories/Strings/Hex.v]: Because `a` is also not a `bool`, and it would need to be something like `if a as a' return if a' then bool else _ then true else b`, which is kind-of hard to read, I think.

**Comment by maximedenes** [comment]
Discussion hasn't converged in time for 8.8, but let's try to integrate or close this PR soon.

File: theories/Strings/Hex.v
Comment (JasonGross) [reply theories/Strings/Hex.v]: Extracted as `ascii_to_digit` function.

File: theories/Strings/Binary.v
Comment (JasonGross) [reply theories/Strings/Binary.v]: Changed

File: theories/Strings/Binary.v
Comment (JasonGross) [reply theories/Strings/Binary.v]: Changed

**Comment by JasonGross** [comment]
I've rebased on master and:
- updated to `{Binary,Octal,Hex}String.v`
- shortened the function names as suggested
- moved the quoted definitions and lemmas to a `Module Raw`, and shortened and unquoted them

> Globally, the current indentation of these files is quite awful.

The indentation is what's given by PG's autoindenter, except for the `if` statements, where I've removed leading whitespace to line up the `else` cases.  Is there a particular change you'd like to see in the indentation?

Finally,

> Overall, I'm not really in favor of merging this PR, even after revision. But I'm not strongly opposed to it either.

I would still like to see this merged.  I think, in as much as this is considered to be a special use case, it's probably because `String` is not often used in Coq.  As a reference point, many other languages provide built-ins for converting between numbers and hex,oct,bin, and decimal, and most languages with printf/scanf support have ways of converting via `%d`, `%o`, `%x`.

**Review by maximedenes — APPROVED** [review]
The template file looks ok

**Comment by JasonGross** [comment]
Who should be the assignee?

**Comment by JasonGross** [comment]
@herbelin Are you willing to be the assignee on this, as the code owner other than @maximedenes ?

**Comment by herbelin** [comment]
> Are you willing to be the assignee on this, as the code owner other than @maximedenes ?

Yes.

Thanks for the changes. Shall merge on Monday late if no objections by then.

**Comment by JasonGross** [comment]
Thanks!

**Comment by maximedenes** [comment]
> @herbelin Are you willing to be the assignee on this, as the code owner other than @maximedenes ?

Sorry, I should have replied earlier. I'm an owner only for the `doc` files, so I think @herbelin is the natural assignee.

File: theories/Strings/BinaryString.v
Comment (liyishuai) [inline theories/Strings/BinaryString.v]: Does this say that `to_pos "0" = 1`? That's semantically counter-intuitive.
Do we really want to convert a string to `positive`?

Comment (JasonGross) [reply theories/Strings/BinaryString.v]: Perhaps, but there's basis for it in the existing standard library:
```coq
Require Import Coq.PArith.PArith.
Require Import Coq.ZArith.ZArith.

Check eq_refl : Z.to_pos 0 = 1%positive. (* success *)
```

Comment (JasonGross) [reply theories/Strings/BinaryString.v]: This is the standard decision that has to be made: do you provide a total function with default values, or do you return `option`?  In my experience, most things are nicer when either literally everything is inside the `option` monad, or when nothing is inside the `option` monad.  For example, it seems not to painful to use `option nat` everywhere instead of `nat`, but if we have `Nat.add : nat -> nat -> nat` and `Nat.sub : nat -> nat -> option nat`, this would be quite painful.

Comment (liyishuai) [reply theories/Strings/BinaryString.v]: I think we can avoid `option` monad for `X -> X -> X` operators to form a monoid.
However, for converting between different types, if we already know that many conversions are invalid (especially converting from strings), then using `option` seems better, unless we have good support for exceptions.

**Pull request merged**
